access/validator.go

```diff
@@ -259,6 +264,19 @@ func (v *TransactionValidator) checkAddresses(tx *flow.TransactionBody) error {
 	return nil
 }
 
+// every key (account, key index combination) can only be used once for signing
+func (v *TransactionValidator) checkSignatureDuplications(tx *flow.TransactionBody) error {
+	observedSigs := make(map[string]bool)
+	for _, sig := range append(tx.PayloadSignatures, tx.EnvelopeSignatures...) {
```

Comment:
So if i am signing the payload with a key, i am not allowed to sign the envelope with the same key? I feel like this might break some assumptions?

cc @psiemens

Reply:
@ramtinms's implementation is correct, in my opinion. 

It's redundant for the the same account key sign both the payload and envelope, so I'm comfortable with calling this invalid.

Because of this line, it's always the case that one of them will be skipped: https://github.com/onflow/flow-go/blob/a01acbf3ed2e7edbc4192f604d10f3763a1700a1/fvm/transactionVerifier.go#L110-L112

If the account is the payer, only the envelope signature is counted and the payload signature is ignored.
If the account is an authorizer but not a payer, then only the payload signature is counted.
If the account is neither, then the signature isn't counted at all.

I think we should actually separately validate that all of the transaction signatures match one of the declared signers (PAYER, PROPOSER, AUTHORIZERS). This way we reject transactions that have irrelevant extra data.

Reply:
I think here docs need to be updated, it states: 

`Special case: if an account is both the payer and either a proposer or authorizer, it is only required to sign the envelope.
` https://docs.onflow.org/concepts/accounts-and-keys/#authorization-envelope

can be changed to: 

`Special case: if an account is both the payer and either a proposer or authorizer, it is required to only sign the envelope.
`

Reply:
Thanks Deniz for pointing the documentation update, we will also have a look at it.

Reply:
I think there is a little subtlety that makes the suggested sentence not quite valid after this PR. 

Let's look at an example with an account having 4 keys of 50% weight each. The account submits a transaction of which it is both the payer and authorizer. It signs the envelope using `key_1` and `key_2`, and signs the payload using `key_3` and `key_4`. This transaction is valid because it doesn't use each key more than once, while the account is signing both the envelope and payload. 

We will still get back to this section of the doc and make sure it is clearer.

Reply:
Agreed, I think also rejecting transactions when the payer signed the payload can be a valid solution.

Because it also satisfies "we reject transactions that have irrelevant extra data."

Reply:
Agreed, this should be the direction we will be going towards 👌